include/ignition/math/PiecewiseScalarField3.hh

```diff
@@ -0,0 +1,215 @@
+    ///     ScalarField3T must support stream operator overload;
+    ///   - for PiecewiseScalarField3::Minimum to be callable, ScalarField3T
+    ///     must implement a
+    ///     ScalarT Minimum(const Region3<ScalarT> &, Vector3<ScalarT> &)
+    ///     method that computes its minimum in the given region and returns
+    ///     an argument value that yields said minimum.
+    /// \tparam ScalarT a numeric type for which std::numeric_limits<> traits
+    ///   have been specialized.
+    ///
+    /// ## Example
+    ///
+    /// \snippet examples/piecewise_scalar_field3_example.cc complete
+    template<typename ScalarField3T, typename ScalarT>
+    class PiecewiseScalarField3
+    {
+      /// \brief A scalar field P in R^3 and
+      /// the region R in which it is defined
+      public: struct Piece {
+        Region3<ScalarT> region;
+        ScalarField3T field;
+      };
+
+      /// \brief Constructor
+      public: PiecewiseScalarField3() = default;
+
+      /// \brief Constructor
+      /// \param[in] _pieces scalar fields Pn and the regions Rn
+      ///   in which these are defined. Regions may not overlap.
+      public: explicit PiecewiseScalarField3(std::vector<Piece> _pieces)
+      : pieces(std::move(_pieces))
+      {
+        for (size_t i = 0; i < pieces.size(); ++i)
+        {
+          if (pieces[i].region.Empty())
+          {
+            std::cerr << "Region #" << i << " (" << pieces[i].region
+                      << ") in piecewise scalar field definition is empty."
+                      << std::endl;
+          }
+          for (size_t j = i + 1; j < pieces.size(); ++j)
+          {
+            if (pieces[i].region.Intersects(pieces[j].region))
+            {
+              std::cerr << "Detected overlap between regions in "
+                        << "piecewise scalar field definition: "
+                        << "region #" << i << " (" << pieces[i].region
+                        << ") overlaps with region #" << j << " ("
+                        << pieces[j].region << "). Region #" << i
+                        << " will take precedence when overlapping."
+                        << std::endl;
+            }
+          }
+        }
+      }
+
+      /// \brief Define piecewise scalar field as `_field` throughout R^3 space
+      public: static PiecewiseScalarField3 Throughout(ScalarField3T _field)
+      {
+        return PiecewiseScalarField3<ScalarField3T, ScalarT>({
+            {Region3<ScalarT>::Unbounded, std::move(_field)}});
+      }
+
+      /// \brief Evaluate the piecewise scalar field at `_p`
+      /// \param[in] _p piecewise scalar field argument
+      /// \return the result of evaluating `F(_p)`, or NaN
+      ///   if the scalar field is not defined at `_p`
+      public: ScalarT Evaluate(const Vector3<ScalarT> &_p) const
+      {
+        auto it = std::find_if(
+            this->pieces.begin(), this->pieces.end(),
+            [&](const Piece &piece) {
```

Comment:
```suggestion
            [&](const Piece &piece)
            {
```

Reply:
Done in 38cbc349121a83d2b94df6328f581fa38aeb127e.

I wonder why `codecheck` didn't catch it. BTW is there a styleguide somewhere?

Reply:
Yeah it's here:

https://ignitionrobotics.org/docs/all/contributing#style-guides

This kind of thing isn't caught by our linters yet, there's some discussion about this here: https://github.com/ignitionrobotics/ign-gazebo/issues/914